cxx/gendb.cc

```diff
@@ -0,0 +1,157 @@
+    return sample_class_ancestors(prng, class_name, class_item);
+  } else {
+    // These are noisy relation domains along the path from the latent cleanly-
+    // observed class to the record class.
+    std::string ref_field = *class_path_start;
+
+    // If the reference field isn't populated, sample a value from a CRP and
+    // add it to reference_values.
+    std::string ref_class =
+        std::get<ClassVar>(
+            schema.classes.at(class_name).vars.at(ref_field).spec)
+            .class_name;
+    if (!reference_values[class_name].contains(class_item)) {
+      sample_and_incorporate_reference(prng, class_name, class_item, ref_field,
+                                       ref_class);
+    }
+    std::vector<int> items = sample_entities_relation(
+        prng, ref_class, ++class_path_start, class_path_end,
+        reference_values[class_name][class_item][ref_field]);
+    items.push_back(class_item);
+    return items;
+  }
+}
+
+void GenDB::sample_and_incorporate_reference(std::mt19937* prng,
+                                             const std::string& class_name,
+                                             int class_item,
+                                             const std::string& ref_field,
+                                             const std::string& ref_class) {
+  int new_val = domain_crps[ref_class].sample(prng);
+
+  // Generate a unique ID for the sample and incorporate it into the
+  // domain CRP.
+  std::stringstream new_id_str;
+  new_id_str << class_name << class_item << ref_field;
+  int new_id = std::hash<std::string>{}(new_id_str.str());
+  reference_values[class_name][class_item][ref_field] = new_val;
+  domain_crps[ref_class].incorporate(new_id, new_val);
+}
+
+// Recursively incorporates samples into base relations.
+void GenDB::incorporate_query_relation(std::mt19937* prng,
+                                       const std::string& query_rel_name,
+                                       const T_items& items,
+                                       const ObservationVariant& value) {
+  RelationVariant query_rel = hirm->get_relation(query_rel_name);
+  T_items base_items = std::visit(
+      [&](auto nr) {
+        using T = typename std::remove_pointer_t<decltype(nr)>::ValueType;
+        auto noisy_rel = reinterpret_cast<NoisyRelation<T>*>(nr);
+        return noisy_rel->get_base_items(items);
+      },
+      query_rel);
+
+  T_noisy_relation t_query_rel =
+      std::get<T_noisy_relation>(hirm->schema.at(query_rel_name));
+  bool base_contains_items = std::visit(
+      [&](auto rel) { return rel->get_data().contains(base_items); },
+      hirm->get_relation(t_query_rel.base_relation));
+  if (!base_contains_items) {
+    hirm->sample_and_incorporate_relation(prng, t_query_rel.base_relation,
+                                          base_items);
+  }
+  hirm->incorporate(prng, query_rel_name, items, value);
+}
+
+// Generates a vector of items from the clean relation domains, with the
```

Comment:
This is called whenever class_path.size() == 1, and as I commented on the Model 7 design doc, when record_class_is_clean = False, you can have noisy relations in this case.

Reply:
Sorry, I meant to respond on the doc -- this works for noisy relations too where class_path.size() == 1, since in that case the noisy relation domains/items are the same as those for the clean relation. I put "clean relation" in the method name in the doc, since the most common case is that the class ancestors are items of a clean relation (I updated the method name so now it doesn't refer to "clean relation").